File: loadtest/ws-load-test.js
// k6 WebSocket load test for the GIGGO real-time tracking socket (WBS P5.10).
//
// Ramps to 200 concurrent STOMP-over-WebSocket clients, each of which:
//   1. opens the WebSocket, 2. sends a JWT-authenticated STOMP CONNECT,
//   3. subscribes to a job's location topic, 4. holds, then disconnects.
//
// Run:  k6 run loadtest/ws-load-test.js
//       BASE_URL=http://localhost:8081 WS_URL=ws://localhost:8081 k6 run loadtest/ws-load-test.js
import ws from 'k6/ws';
import http from 'k6/http';
import { check } from 'k6';
import { Counter } from 'k6/metrics';

const BASE = __ENV.BASE_URL || 'http://localhost:8081';
const WS = (__ENV.WS_URL || 'ws://localhost:8081') + '/ws';
const NUL = String.fromCharCode(0); // STOMP frame terminator (null byte)

export const options = {
  scenarios: {
    ramp: {
      executor: 'ramping-vus',
      startVUs: 0,
      stages: [
        { duration: '10s', target: 200 }, // ramp up
        { duration: '20s', target: 200 }, // hold 200 concurrent
        { duration: '5s', target: 0 },    // ramp down
      ],
    },
  },
  thresholds: {
    ws_connecting: ['p(95)<1000'], // 95% of handshakes under 1s
    checks: ['rate>0.99'],         // >99% of connections upgrade + STOMP-connect
  },
};

const stompConnected = new Counter('stomp_connected');

export function setup() {
  const res = http.post(
    `${BASE}/api/v1/auth/login`,
    JSON.stringify({ email: 'demo.customer@giggo.lk', password: 'Provider@123' }),
    { headers: { 'Content-Type': 'application/json' } },
  );
  return { token: res.json('data.accessToken') };
}

function frame(command, headers, body = '') {
  let h = '';
  for (const k in headers) h += `${k}:${headers[k]}\n`;
  return `${command}\n${h}\n${body}${NUL}`;
}

export default function (data) {
  const res = ws.connect(WS, {}, function (socket) {
    socket.on('open', () => {
      socket.send(frame('CONNECT', {
        'accept-version': '1.2',
        host: 'localhost',
        Authorization: `Bearer ${data.token}`,
      }));
    });
    socket.on('message', (msg) => {
      if (msg.startsWith('CONNECTED')) {
        stompConnected.add(1);
        socket.send(frame('SUBSCRIBE', { id: 'sub-0', destination: '/topic/jobs/loadtest/location' }));
        socket.setTimeout(() => socket.close(), 3000);
      }
    });
  });
  check(res, { 'ws upgraded (101)': (r) => r && r.status === 101 });
}

File: loadtest/ws_load_test.py
"""Runnable WebSocket load test for the GIGGO tracking socket (WBS P5.10).

Opens N concurrent STOMP-over-WebSocket clients against the backend. Each client:
  1. opens the WebSocket, 2. sends a JWT-authenticated STOMP CONNECT,
  3. waits for CONNECTED, 4. subscribes to a job's location topic, 5. holds.
Reports the number connected, connect-time percentiles and any failures.

Usage:  py loadtest/ws_load_test.py [N] [BASE_URL]
Example: py loadtest/ws_load_test.py 200 http://localhost:8081
Requires: pip install websockets
"""

import asyncio
import json
import sys
import time
import urllib.request
from collections import Counter

N = int(sys.argv[1]) if len(sys.argv) > 1 else 200
BASE = sys.argv[2] if len(sys.argv) > 2 else "http://localhost:8081"
WS = BASE.replace("http", "ws") + "/ws"
HOLD_SECONDS = 3.0
NUL = "\x00"

import websockets  # noqa: E402  (after arg parsing so --help is cheap)


def login() -> str:
    body = json.dumps({"email": "demo.customer@giggo.lk", "password": "Provider@123"}).encode()
    req = urllib.request.Request(
        f"{BASE}/api/v1/auth/login", data=body, headers={"Content-Type": "application/json"}
    )
    return json.loads(urllib.request.urlopen(req).read())["data"]["accessToken"]


def frame(command: str, headers: dict, body: str = "") -> str:
    h = "".join(f"{k}:{v}\n" for k, v in headers.items())
    return f"{command}\n{h}\n{body}{NUL}"


async def one_client(i: int, token: str, results: list) -> None:
    start = time.monotonic()
    try:
        async with websockets.connect(WS, open_timeout=15, ping_interval=None) as ws:
            await ws.send(frame("CONNECT", {
                "accept-version": "1.2", "host": "localhost", "Authorization": f"Bearer {token}",
            }))
            reply = await asyncio.wait_for(ws.recv(), timeout=10)
            if not reply.startswith("CONNECTED"):
                results.append(("not_connected", time.monotonic() - start))
                return
            await ws.send(frame("SUBSCRIBE", {"id": f"sub-{i}", "destination": "/topic/jobs/loadtest/location"}))
            results.append(("ok", time.monotonic() - start))
            await asyncio.sleep(HOLD_SECONDS)
    except Exception as exc:  # noqa: BLE001 — record the failure class
        results.append((type(exc).__name__, time.monotonic() - start))


async def main() -> None:
    print(f"Load test: {N} concurrent WebSocket clients -> {WS}")
    token = login()
    results: list = []
    started = time.monotonic()
    await asyncio.gather(*[one_client(i, token, results) for i in range(N)])
    wall = time.monotonic() - started

    ok = sorted(d for s, d in results if s == "ok")
    failures = [s for s, _ in results if s != "ok"]

    def pctl(p: float) -> float:
        return ok[min(len(ok) - 1, int(len(ok) * p))] * 1000 if ok else 0.0

    print(f"connected: {len(ok)}/{N}   failed: {len(failures)}")
    if ok:
        print(f"connect time ms — min {ok[0]*1000:.0f} | p50 {pctl(0.5):.0f} | "
              f"p95 {pctl(0.95):.0f} | max {ok[-1]*1000:.0f}")
    print(f"wall time: {wall:.1f}s")
    if failures:
        print("failures:", dict(Counter(failures)))


if __name__ == "__main__":
    asyncio.run(main())
